File: web/src/types/graph.ts
// Canonical frontend types. These mirror the Go structs in
// internal/model/{graph,plan,connection,errors}.go and internal/simulate/cascade.go.
// Field names MUST stay in sync with the JSON tags on those structs — do not rename.
// This is the frozen shared contract for all feature agents (DESIGN_PLAN §2.3, §8).

export type Engine = 'postgres' | 'mongodb' | 'sqlite';

export interface ConnectionConfig {
  engine: Engine | '';
  dsn?: string;
  host?: string;
  port?: number;
  user?: string;
  password?: string;
  database?: string;
  filePath?: string;
  label?: string;
}

export interface ConnMeta {
  id: string;
  engine: string;
  state: string;
  config: ConnectionConfig;
  openedAt: string;
}

export interface FKRef {
  table: string;
  column: string;
  onDelete: string;
  onUpdate: string;
  inferred: boolean;
  confidence?: number;
}

export interface Column {
  name: string;
  type: string;
  nullable: boolean;
  isPk: boolean;
  isFk: boolean;
  fkRef?: FKRef;
  indexes?: string[];
  default?: string;
}

export interface Node {
  id: string;
  schema: string;
  label: string;
  kind: string;
  columns: Column[];
  rowCount: number;
  sizeBytes: number;
  tenant?: string;
}

export interface Link {
  source: string;
  target: string;
  type: string;
  cascade: boolean;
  viaColumn: string;
  inferred: boolean;
  confidence?: number;
}

export interface Stats {
  nodeCount: number;
  linkCount: number;
  cascadeChainDepth: number;
  columnCount: number;
}

export interface Warning {
  code: string;
  message: string;
}

export interface GraphModel {
  engine: string;
  database: string;
  schemas: string[];
  nodes: Node[];
  links: Link[];
  stats: Stats;
  warnings?: Warning[];
}

export interface TableStats {
  nodeId: string;
  rowCount: number;
  deadTuples?: number;
  sizeBytes: number;
  inserts?: number;
  updates?: number;
  deletes?: number;
  lastVacuum?: string;
  lastAutovacuum?: string;
  lastAnalyze?: string;
}

export interface CascadeStep {
  nodeId: string;
  via: string;
  viaColumn: string;
  depth: number;
}

export interface CascadeResult {
  root: string;
  affected: CascadeStep[];
}

export interface ExplainNode {
  table: string;
  op: string;
  cost: number;
}

export interface QueryPlan {
  query: string;
  nodes: ExplainNode[];
}

export type CrudOp = 'insert' | 'update' | 'delete';

export interface CrudFlowStep {
  nodeId: string;
  order: number;
  action: 'insert' | 'update' | 'delete' | 'check';
  via?: string;
  viaColumn?: string;
}

export interface CrudFlowResult {
  operation: CrudOp;
  root: string;
  steps: CrudFlowStep[];
}

export interface DockerContainer {
  id: string;
  name: string;
  image: string;
  engine: string;
  ports: string[];
}

export type ErrorCode =
  | 'CONN_INVALID_DSN'
  | 'CONN_AUTH_FAILED'
  | 'CONN_UNREACHABLE'
  | 'CONN_SUPERUSER_REJECTED'
  | 'CONN_NOT_FOUND'
  | 'CONN_TIMEOUT'
  | 'CONN_ALREADY_CLOSED'
  | 'SCHEMA_INTROSPECT_FAILED'
  | 'SCHEMA_PERMISSION_DENIED'
  | 'SCHEMA_NOT_FOUND'
  | 'SCHEMA_TOO_LARGE'
  | 'EXPLAIN_INVALID_SQL'
  | 'EXPLAIN_NOT_SUPPORTED'
  | 'EXPLAIN_TIMEOUT'
  | 'ADAPTER_NOT_SUPPORTED'
  | 'ADAPTER_OPERATION_NOT_SUPPORTED'
  | 'ADAPTER_READONLY_VIOLATION'
  | 'INTERNAL'
  | 'BAD_REQUEST'
  | 'RATE_LIMITED';

// Imperative handle the graph registers so the ActionBar can drive animations without prop-drilling the D3 sim.
export interface GraphAnimator {
  animateCascade: (r: CascadeResult) => void;
  animateQueryPath: (p: QueryPlan) => void;
  animateCrud: (r: CrudFlowResult) => void;
  pulse: (nodeId: string) => void;
  clearHighlights: () => void;
}

File: internal/model/graph.go
// Package model defines the canonical, engine-agnostic data shapes that every
// adapter normalizes into. GraphModel is the contract between adapters and the
// frontend — see DESIGN_PLAN §2.3.
package model

// GraphModel is the unified schema representation produced by every adapter.
type GraphModel struct {
	Engine   string    `json:"engine"` // "postgres" | "mongodb" | ...
	Database string    `json:"database"`
	Schemas  []string  `json:"schemas"` // e.g. ["public", "audit"] for Postgres
	Nodes    []Node    `json:"nodes"`
	Links    []Link    `json:"links"`
	Stats    Stats     `json:"stats"`
	Warnings []Warning `json:"warnings,omitempty"`
}

// Node is a table, collection, or view in the graph.
type Node struct {
	ID        string   `json:"id"`     // "<schema>.<name>" — see §11
	Schema    string   `json:"schema"` // "public" for Postgres; "" for engines w/o schemas
	Label     string   `json:"label"`  // display name (table name, no schema prefix)
	Kind      string   `json:"kind"`   // "table" | "collection" | "view" | "index_pattern"
	Columns   []Column `json:"columns"`
	RowCount  int64    `json:"rowCount"`
	SizeBytes int64    `json:"sizeBytes"`
	Tenant    string   `json:"tenant,omitempty"` // Phase 2
}

// Column is a single field on a Node.
type Column struct {
	Name       string   `json:"name"`
	Type       string   `json:"type"` // canonical, see §2.4
	Nullable   bool     `json:"nullable"`
	IsPK       bool     `json:"isPk"`
	IsFK       bool     `json:"isFk"`
	FKRef      *FKRef   `json:"fkRef,omitempty"`
	Indexes    []string `json:"indexes,omitempty"`
	DefaultVal string   `json:"default,omitempty"`
}

// FKRef describes a foreign-key relationship from a Column to another node.
type FKRef struct {
	Table      string  `json:"table"` // referenced node ID
	Column     string  `json:"column"`
	OnDelete   string  `json:"onDelete"` // "CASCADE" | "SET NULL" | "RESTRICT" | "NO ACTION"
	OnUpdate   string  `json:"onUpdate"`
	Inferred   bool    `json:"inferred"` // true for NoSQL heuristic
	Confidence float64 `json:"confidence,omitempty"`
}

// Link is an edge between two nodes, derived from FK relationships.
type Link struct {
	Source     string  `json:"source"` // node ID
	Target     string  `json:"target"` // node ID
	Type       string  `json:"type"`   // "1:1" | "1:N" | "M:N"
	Cascade    bool    `json:"cascade"`
	ViaColumn  string  `json:"viaColumn"`
	Inferred   bool    `json:"inferred"`
	Confidence float64 `json:"confidence,omitempty"`
}

// Stats summarizes the graph.
type Stats struct {
	NodeCount         int `json:"nodeCount"`
	LinkCount         int `json:"linkCount"`
	CascadeChainDepth int `json:"cascadeChainDepth"`
	ColumnCount       int `json:"columnCount"`
}

// Warning is a non-fatal advisory attached to a GraphModel.
type Warning struct {
	Code    string `json:"code"` // "NODE_COUNT_HIGH" | "SCHEMA_FILTERED" | ...
	Message string `json:"message"`
}

// Warning codes used in GraphModel.Warnings.
const (
	WarnNodeCountHigh    = "NODE_COUNT_HIGH"
	WarnSchemaFiltered   = "SCHEMA_FILTERED"
	WarnColumnsTruncated = "COLUMNS_TRUNCATED"
)

// Node kinds.
const (
	KindTable        = "table"
	KindCollection   = "collection"
	KindView         = "view"
	KindIndexPattern = "index_pattern"
)

// Link cardinality types.
const (
	Link1To1 = "1:1"
	Link1ToN = "1:N"
	LinkMToN = "M:N"
)

// FK referential actions.
const (
	OnCascade  = "CASCADE"
	OnSetNull  = "SET NULL"
	OnRestrict = "RESTRICT"
	OnNoAction = "NO ACTION"
)

// Canonical column types — every adapter maps engine-specific types to one of
// these (§2.4). The original engine type is preserved as a suffix, e.g.
// "uuid (uuid)", "string (varchar(255))".
const (
	TypeString    = "string"
	TypeText      = "text"
	TypeInteger   = "integer"
	TypeBigint    = "bigint"
	TypeDecimal   = "decimal"
	TypeFloat     = "float"
	TypeBoolean   = "boolean"
	TypeTimestamp = "timestamp"
	TypeDate      = "date"
	TypeUUID      = "uuid"
	TypeJSON      = "json"
	TypeBinary    = "binary"
	TypeArray     = "array"
	TypeEnum      = "enum"
	TypeGeometry  = "geometry"
	TypeUnknown   = "unknown"
)

// TableStats holds per-table statistics surfaced at
// GET /api/connections/:id/tables/:nodeId/stats.
type TableStats struct {
	NodeID         string `json:"nodeId"`
	RowCount       int64  `json:"rowCount"`
	DeadTuples     int64  `json:"deadTuples,omitempty"`
	SizeBytes      int64  `json:"sizeBytes"`
	Inserts        int64  `json:"inserts,omitempty"`
	Updates        int64  `json:"updates,omitempty"`
	Deletes        int64  `json:"deletes,omitempty"`
	SeqScans       int64  `json:"seqScans,omitempty"`
	IdxScans       int64  `json:"idxScans,omitempty"`
	LastVacuum     string `json:"lastVacuum,omitempty"`
	LastAutovacuum string `json:"lastAutovacuum,omitempty"`
	LastAnalyze    string `json:"lastAnalyze,omitempty"`
}
